src/Build/Logging/BinaryLogger/ProjectImportsCollector.cs

```diff
@@ -0,0 +1,159 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Threading.Tasks;
+
+namespace Microsoft.Build.Logging
+{
+    /// <summary>
+    /// Creates a zip archive with all the .csproj and .targets encountered during the build.
+    /// The internal .zip file structure matches closely the layout of the original sources on disk.
+    /// The .zip file can be used to correlate the file names and positions in the build log file with the
+    /// actual sources.
+    /// </summary>
+    internal class ProjectImportsCollector
+    {
+        private FileStream _fileStream;
+        private ZipArchive _zipArchive;
+
+        public string ArchiveFilePath { get; set; }
+
+        /// <summary>
+        /// Avoid visiting each file more than once.
+        /// </summary>
+        private readonly HashSet<string> _processedFiles = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Extensions to exclude from the .buildsources.zip archive
+        /// </summary>
+        private static readonly HashSet<string> _extensionsToExclude = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            ".sln",
+            ".cs",
+            ".h",
+            ".cpp",
+            ".fs",
+            ".vb"
+        };
+
+        // this will form a chain of file write tasks, running sequentially on a background thread
+        private Task _currentTask = Task.CompletedTask;
+
+        public ProjectImportsCollector(string logFilePath, string sourcesArchiveExtension = ".buildsources.zip")
+        {
+            ArchiveFilePath = Path.ChangeExtension(logFilePath, sourcesArchiveExtension);
+
+            try
+            {
+                _fileStream = new FileStream(ArchiveFilePath, FileMode.Create, FileAccess.ReadWrite, FileShare.Delete);
+                _zipArchive = new ZipArchive(_fileStream, ZipArchiveMode.Create);
+            }
+            catch
+            {
+                // For some reason we weren't able to create a file for the archive.
+                // Disable the file collector.
+                _fileStream = null;
+                _zipArchive = null;
+            }
+        }
+
+        public void AddFile(string filePath)
+        {
+            if (filePath == null || _fileStream == null)
+            {
+                return;
+            }
+
+            lock (_fileStream)
+            {
+                // enqueue the task to add a file and return quickly
+                // to avoid holding up the current thread
+                _currentTask = _currentTask.ContinueWith(t =>
+                {
+                    try
+                    {
+                        AddFileCore(filePath);
```

Comment:
Yeah I was thinking about that. We could issue an error and send it to the log, but it's on a different thread... The catch block has to be there because if an exception bubbles up from a thread like this nobody will catch it and it will tear down the process. Since it's non-essential functionality (worst case we lose a file) I thought it's OK to be on the safer side and just hope it never gets hit.

I have a system where I regularly exercise MSBuild and monitor all first-chance exceptions. If I notice anything really happening there we can think how to mitigate. I think it's OK for now.